manifests/prepare/puppet_config.pp

```diff
@@ -2,37 +2,59 @@
-      'certificate_expire_warning', 'couchdb_url', 'dbadapter', 'dbconnections', 'dblocation', 'dbmigrate', 'dbname',
-      'dbpassword', 'dbport', 'dbserver', 'dbsocket', 'dbuser', 'dynamicfacts', 'http_compression', 'httplog',
-      'ignoreimport', 'immutable_node_data', 'inventory_port', 'inventory_server', 'inventory_terminus',
-      'legacy_query_parameter_serialization', 'listen', 'localconfig', 'manifestdir', 'masterlog', 'parser',
-      'preview_outputdir', 'puppetport', 'queue_source', 'queue_type', 'rails_loglevel', 'railslog',
-      'report_serialization_format', 'reportfrom', 'rrddir', 'rrdinterval', 'sendmail', 'smtphelo', 'smtpport',
-      'smtpserver', 'stringify_facts', 'tagmap', 'templatedir', 'thin_storeconfigs', 'trusted_node_data', 'zlib',
-      # Deprecated for global config
-      'config_version', 'manifest', 'modulepath',
-      # Settings that should be reset to defaults
-      'disable_warnings', 'vardir', 'rundir', 'libdir', 'confdir', 'ssldir', 'classfile'].each |$setting| {
-      ini_setting { "${section}/${setting}":
-        ensure  => absent,
-        section => $section,
-        setting => $setting,
-        path    => $puppetconf,
-        require => File[$puppetconf],
+
+    if versioncmp("${::clientversion}", '4.0.0') < 0 {
+      $_removedSettings = [# Removed settings
+        'allow_variables_with_dashes', 'async_storeconfigs', 'binder', 'catalog_format', 'certdnsnames',
+        'certificate_expire_warning', 'couchdb_url', 'dbadapter', 'dbconnections', 'dblocation', 'dbmigrate', 'dbname',
+        'dbpassword', 'dbport', 'dbserver', 'dbsocket', 'dbuser', 'dynamicfacts', 'http_compression', 'httplog',
+        'ignoreimport', 'immutable_node_data', 'inventory_port', 'inventory_server', 'inventory_terminus',
+        'legacy_query_parameter_serialization', 'listen', 'localconfig', 'manifestdir', 'masterlog', 'parser',
+        'preview_outputdir', 'puppetport', 'queue_source', 'queue_type', 'rails_loglevel', 'railslog',
+        'report_serialization_format', 'reportfrom', 'rrddir', 'rrdinterval', 'sendmail', 'smtphelo', 'smtpport',
+        'smtpserver', 'stringify_facts', 'tagmap', 'templatedir', 'thin_storeconfigs', 'trusted_node_data', 'zlib',
+        # Deprecated for global config
+        'config_version', 'manifest', 'modulepath',
+        # Settings that should be reset to defaults
+        'disable_warnings', 'vardir', 'rundir', 'libdir', 'confdir', 'ssldir', 'classfile']
+    } else {
+      $_removedSettings = []
+    }
+
+    # When upgrading to 1.4.x or later remove pluginsync
+    if (($package_version == undef and $old_packages) or (versioncmp($package_version, '1.4.0') >= 0))
+        and !defined(Ini_setting["${section}/pluginsync"]) {
+      $removedSettings = $_removedSettings + ['pluginsync']
+    } else {
+      $removedSettings = $_removedSettings
+    }
+
+    if $removedSettings {
```

Comment:
Yeah, we can probably remove the if. As for my comment about the `.each`, I'm just trying to figure out where `no implicit conversion of nil into String` is coming from.